optuna/storages/grpc/_server.py

```diff
@@ -0,0 +1,474 @@
+from __future__ import annotations
+
+from concurrent.futures import ThreadPoolExecutor
+from datetime import datetime
+import json
+import threading
+
+from optuna.distributions import distribution_to_json
+from optuna.distributions import json_to_distribution
+from optuna.exceptions import DuplicatedStudyError
+from optuna.storages import BaseStorage
+from optuna.storages.grpc._grpc_imports import _imports
+from optuna.study._study_direction import StudyDirection
+from optuna.trial._frozen import FrozenTrial
+from optuna.trial._state import TrialState
+
+
+if _imports.is_successful():
+    from optuna.storages.grpc._grpc_imports import api_pb2
+    from optuna.storages.grpc._grpc_imports import api_pb2_grpc
+    from optuna.storages.grpc._grpc_imports import grpc
+    from optuna.storages.grpc._grpc_imports import StorageServiceServicer
+else:
+
+    class StorageServiceServicer:  # type: ignore
+        pass
+
+
+DATETIME_FORMAT = "%Y-%m-%d %H:%M:%S.%f"
+
+
+class OptunaStorageProxyService(StorageServiceServicer):
+    def __init__(self, storage: BaseStorage) -> None:
+        self._backend = storage
+        self._lock = threading.Lock()
+
+    def CreateNewStudy(
+        self,
+        request: api_pb2.CreateNewStudyRequest,
+        context: grpc.ServicerContext,
+    ) -> api_pb2.CreateNewStudyReply:
+        directions = [
+            StudyDirection.MINIMIZE if d == api_pb2.MINIMIZE else StudyDirection.MAXIMIZE
+            for d in request.directions
+        ]
+        study_name = request.study_name
+
+        try:
+            study_id = self._backend.create_new_study(directions=directions, study_name=study_name)
+        except DuplicatedStudyError as e:
+            context.abort(code=grpc.StatusCode.ALREADY_EXISTS, details=str(e))
+        return api_pb2.CreateNewStudyReply(study_id=study_id)
+
+    def DeleteStudy(
+        self,
+        request: api_pb2.DeleteStudyRequest,
+        context: grpc.ServicerContext,
+    ) -> api_pb2.DeleteStudyReply:
+        study_id = request.study_id
+        try:
+            self._backend.delete_study(study_id)
+        except KeyError as e:
+            context.abort(code=grpc.StatusCode.NOT_FOUND, details=str(e))
```

Comment:
The same approach as in `optuna/storages/grpc/_client.py` may be applicable here as well.

Reply:
In the server side code, we need to treat with the different types of errors and set different types of grpc status codes according to them. So, I think it would be better to leave them as they are in the server side. (BTW, thanks for the nice suggestion in the client side!)